### PR TITLE
KEYCLOAK-2054 - Allow to configure proxy for auth-server requests in adapters.

### PR DESCRIPTION
**Pull request by thomasdarimont**
Previously the adapter configuration did not support to specify a proxy
to use for auth-server requests issued via the Apache HTTP Client used by
Keycloak. 
This made it very difficult to connect an Application with that was required to use a proxy.

Introduced 4 new attributes to the adapter configuration
(e.g. keycloak.json) which makes it possible to
configure a proxy to be used for auth-server requests.
- proxy-enabled: Whether to use Proxy configuration, defaults to false
- proxy-host: The Proxy Host configuration, defaults to null
- proxy-port: The Proxy Port configuration, defaults to null
- proxy-protocol-scheme: The Proxy Protocol configuration, defaults to
  http

Added documentation.

### CODE REVIEW

**Comment by mstruk** [comment]
We would also have to update the Wildfly Keycloak subsystem as it should
mirror the adapter config options available via keycloak.json. Not sure if
this would have to be part of the same PR, but probably yes.

**Comment by thomasdarimont** [comment]
Hello,

with `wildfly-subsystem`you mean this? https://github.com/keycloak/keycloak/tree/master/wildfly/server-subsystem

What should be changed here?

Cheers,
Thomas

**Comment by mstruk** [comment]
It's the adapter subsystems - several places where we do this:

https://github.com/keycloak/keycloak/tree/master/adapters/oidc/wildfly/wildfly-subsystem

https://github.com/keycloak/keycloak/tree/master/adapters/oidc/wildfly/wf8-subsystem

https://github.com/keycloak/keycloak/tree/master/adapters/oidc/as7-eap6/as7-subsystem

I can take care of this part.

**Comment by thomasdarimont** [comment]
Would be great :) I think this can then be done with a separate PR.

But we should first make sure that ther suggested way for configuring the proxy for the auth-server is sound ;-)

The suggested appoach is working fine in our tests...

**Comment by stianst** [comment]
Is it not clunky to have 3 attributes? Could we not just have proxy url?

**Comment by thomasdarimont** [comment]
Sure I can reduce this to one url ... how about `auth-server-proxy-url`?

**Comment by thomasdarimont** [comment]
Updated PR accordingly.

**Comment by stianst** [comment]
At the risk of being picky wouldn't just proxy-url be simpler and also more clear? auth-server-proxy-url makes me think it's a proxy provided by the auth server, rather than a proxy the client has to go through to get outbound connection.

**Comment by thomasdarimont** [comment]
My intent is to explicitly communicate that all auth server requests go through the configured proxy.
Because keycloak-server.json doesn't support comments, I thought it would be better to be explicit here.

(a shorter name is fine by me :)

by the way - I think it would be good to enable support for comments in the "keycloak-server" or even Adpater configuration parser, which could help give in-place rationale for particular config settings
```
// to allow C/C++ style comments in JSON (non-standard, disabled by default)
// (note: with Jackson 2.5, there is also `mapper.enable(feature)` / `mapper.disable(feature)`)
mapper.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
```

**Comment by stianst** [comment]
I think it's assumed that by adding a proxy config the requests will go through that. Maybe, I'm mistaken, but to me at least the current name makes me think it's something to do with the server, not a client side thing.

This is for adapter keycloak.json and this option is added manually for users, so not sure how allowing comments helps. Further, that's what documentation is for. So -1 to comments in json file from me.

BTW keycloak-server.json will be removed at some point and we'll make server configuration in standalone.xml.

**Comment by thomasdarimont** [comment]
As said - I can change it to proxy-url, if that's more clear.

Regarding the comment support in config files I have to disagree - the documentation is fine for explaining possible supported options, bit I want to be able to document /why/ a certain settings is configured this way - sure having the config in VCS helps here - but it is more convenient to have the info in the form of comments.

Thanks for sharing the hint regarding the keycloak-server.json to standalone.xml transition.
I guess the plan is to be able to configure Keycloak via wildfly-cli then, could be pretty handy.

**Comment by stianst** [comment]
IMO it's more clear, but I'm only one person, so maybe I'm the only one that would be confused ;)

I can't see how it would be added to the keycloak.json file though. keycloak.json can be fully crafted by the user, or partially crafted by the admin console (installation tab), but that doesn't included fields like this. Or maybe I'm misunderstanding what you are proposing. Oh, wait I think I just got it... You as a user of Keycloak want to add a comment in your application code, right? If so that makes sense, create a JIRA and send a PR if you want.

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java [hunk 1]
```diff
@@ -287,16 +289,20 @@ public HttpClient build() {
                 cm = new SingleClientConnManager(registry);
             }
             BasicHttpParams params = new BasicHttpParams();
-            if (socketTimeout > -1)
-            {
+
+            if (proxyHost != null) {
+                params.setParameter(ConnRoutePNames.DEFAULT_PROXY, proxyHost);
```
Comment (stianst) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java]: Looks like this is deprecated - https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/conn/params/ConnRoutePNames.html#DEFAULT_PROXY

Comment (thomasdarimont) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java]: Yes I know - unfortunately they didn't provide an alternative and it works with this version.
I didn't want to upgrade or introduce a new HTTPClient Version for this.

Comment (stianst) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java]: What about the following from section 2.8 in https://hc.apache.org/httpcomponents-client-ga/tutorial/html/connmgmt.html:
`
HttpHost proxy = new HttpHost("someproxy", 8080);
DefaultProxyRoutePlanner routePlanner = new DefaultProxyRoutePlanner(proxy);
CloseableHttpClient httpclient = HttpClients.custom()
        .setRoutePlanner(routePlanner)
        .build();
`

Comment (thomasdarimont) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java]: Thanks - I'll give it a try and adjust the PR if it works.

Comment (thomasdarimont) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java]: Just gave it a quick spin - turns out that I'd probably need to port the whole HTTP Client config to the new HttpClientBuilder API.

I think that is a bit out of scope for this PR. I'd rather go with the small change (using the deprectaed but working API) and port the HTTP config in a separate PR. wdyt?

**Comment by stianst** [comment]
Sorry for delay in response. Let's go with the deprecated config for now. If you can rename auth-server-proxy-url to just proxy-url then we can get this merged. Thanks.

**Comment by thomasdarimont** [comment]
Thanks for the feedback @stianst I'll update the PR.

**Comment by thomasdarimont** [comment]
PR is updated.

**Comment by stianst** [comment]
Thanks

**Comment by schtrym** [comment]
Is it possible to configure the proxy through the standalone.xml?

**Comment by thomasdarimont** [comment]
Hi @schtrym,

yes it is. I don't have an example at hand but, you could just configure a proxy via the example jboss-cli script and look at the resulting `standalone.xml` / `standalone-ha.xml`.

Cheers,
Thomas

**Comment by schtrym** [comment]
Thanks for the reply Thomas.

i tried adding the following in the `standalone.xml`

```
		<property name="http.proxyHost" value="localhost"/>
		<property name="http.proxyPort" value="3129"/>
		<property name="https.proxyHost" value="localhost"/>
		<property name="https.proxyPort" value="3129"/>
    </system-properties>
```

or in the standalone.conf.bat
```
set "JAVA_OPTS=%JAVA_OPTS% -Dhttp.proxyHost=localhost -Dhttp.proxyPort=3129 -Dhttps.proxyHost=localhost -Dhttps.proxyPort=3129"
```

but without success.

As I read this issue, I understand that's a special configuration for the keycloak adapter. no?

**Comment by thomasdarimont** [comment]
The proxy url attribute for the Keycloak adapter is `proxy-url`.

If you'd like to configure the proxy for keycloak in general then have a look at this: https://github.com/keycloak/keycloak/pull/4543

**Comment by schtrym** [comment]
Where do I have to put this attribute `proxy-url` in my standalone.xml?

in the `<subsystem xmlns="urn:jboss:domain:keycloak:1.1">` ?
